contracts/RegistryV1.sol

```diff
@@ -0,0 +1,313 @@
+   * @param rewardsConsumer The consumer entity address
+   */
+  function approveRewardsAgreement(address rewardsConsumer) external {
+    uint256 providerId = _entityToId[msg.sender];
+    uint256 consumerId = _entityToId[rewardsConsumer];
+
+    if (providerId == 0 || consumerId == 0) {
+      revert EntityDoesNotExist(providerId == 0 ? msg.sender : rewardsConsumer);
+    }
+
+    // Create the agreement
+    bytes32 agreementKey = keccak256(abi.encodePacked(providerId, consumerId));
+    _agreements[agreementKey] = true;
+    emit RewardsAgreementApproved(msg.sender, rewardsConsumer);
+  }
+
+  /**
+   * @notice Registers a user as being referred to a rewards agreement
+   * @param user The address of the user being referred
+   * @param userSignature The user's signature authorizing the referral
+   * @param rewardsConsumer The address of the rewards consumer entity
+   * @param rewardsProvider The address of the rewards provider entity
+   * @return success Whether the referral was successfully registered
+   */
+  function registerReferral(
+    address user,
+    bytes memory userSignature,
+    address rewardsConsumer,
+    address rewardsProvider
+  ) public returns (bool success) {
+    // Verify user signature
+    bytes32 hash = keccak256(abi.encodePacked('Confirm referral'));
+    address signer = hash.toEthSignedMessageHash().recover(userSignature);
+    if (user != signer) {
+      emit ReferralFailed(
+        rewardsProvider,
+        rewardsConsumer,
+        user,
+        'InvalidSignature'
+      );
+      return false;
+    }
+
+    uint256 consumerId = _entityToId[rewardsConsumer];
+    uint256 providerId = _entityToId[rewardsProvider];
+
+    if (consumerId == 0 || providerId == 0) {
+      emit ReferralFailed(
+        rewardsProvider,
+        rewardsConsumer,
+        user,
+        'EntityDoesNotExist'
+      );
+      return false;
+    }
+
+    // Check if agreement exists and is approved
+    bytes32 agreementKey = keccak256(abi.encodePacked(providerId, consumerId));
+    if (!_agreements[agreementKey]) {
+      emit ReferralFailed(
+        rewardsProvider,
+        rewardsConsumer,
+        user,
+        'AgreementDoesNotExist'
+      );
+      return false;
+    }
+
+    // Skip if user is already referred to this provider
+    if (_userReferrals[user][providerId] != 0) {
+      emit ReferralFailed(
+        rewardsProvider,
+        rewardsConsumer,
+        user,
+        'UserAlreadyReferred'
+      );
+      return false;
+    }
+
+    // Add referral
+    _userReferrals[user][providerId] = consumerId;
+    emit ReferralRegistered(rewardsProvider, rewardsConsumer, user);
+    return true;
+  }
+
+  /**
+   * @notice Get the consumer address for a given user and provider
+   * @param user The address of the user
+   * @param provider The address of the provider
+   * @return The address of the consumer
+   */
+  function getReferringConsumer(
```

Comment:
i'm not sure if we need this, but in case a registerReferral is unsuccessful and the event emitted is "UserAlreadyReferred", i thought a builder might want an easy way to know which consumer the user is referred by for debugging?